arangod/Pregel/Actor/include/Actor/Actor.h

```diff
@@ -0,0 +1,223 @@
+concept MessageIsVariant = requires(ActorPID pid,
+                                    std::shared_ptr<Runtime> runtime,
+                                    typename A::State state,
+                                    typename A::Message message) {
+  {std::visit(
+      typename A::template Handler<Runtime>{
+          {pid, pid, std::move(std::make_unique<typename A::State>(state)),
+           runtime}},
+      message)};
+};
+template<typename A>
+concept IsInspectable = requires(typename A::Message message,
+                                 typename A::State state) {
+  {fmt::format("{}", message)};
+  {fmt::format("{}", state)};
+};
+};  // namespace
+template<typename Runtime, typename A>
+concept Actorable = IncludesAllActorRelevantTypes<Runtime, A> &&
+    IncludesName<A> && HandlerInheritsFromBaseHandler<Runtime, A> &&
+    MessageIsVariant<Runtime, A> && IsInspectable<A>;
+
+template<typename Runtime, typename Config>
+requires Actorable<Runtime, Config>
+struct Actor : ActorBase {
+  Actor(ActorPID pid, std::shared_ptr<Runtime> runtime,
+        std::unique_ptr<typename Config::State> initialState)
+      : pid(pid), runtime(runtime), state(std::move(initialState)) {}
+
+  Actor(ActorPID pid, std::shared_ptr<Runtime> runtime,
+        std::unique_ptr<typename Config::State> initialState,
+        std::size_t batchSize)
+      : pid(pid),
+        runtime(runtime),
+        state(std::move(initialState)),
+        batchSize(batchSize) {}
+
+  ~Actor() = default;
+
+  auto typeName() -> std::string_view override { return Config::typeName(); };
+
+  void process(ActorPID sender,
+               std::unique_ptr<MessagePayloadBase> msg) override {
+    auto* ptr = msg.release();
+    if (auto* m = dynamic_cast<MessagePayload<typename Config::Message>*>(ptr);
+        m != nullptr) {
+      push(sender, std::move(m->payload));
+      delete m;
+    } else if (auto* n = dynamic_cast<MessagePayload<ActorError>*>(ptr);
+               n != nullptr) {
+      push(sender, std::move(n->payload));
+      delete n;
+    } else {
+      runtime->dispatch(
+          pid, sender,
+          ActorError{UnknownMessage{.sender = sender, .receiver = pid}});
+      delete ptr;
+    }
+    kick();
+  }
+
+  void process(ActorPID sender, velocypack::SharedSlice msg) override {
+    if (auto m =
+            inspection::deserializeWithErrorT<typename Config::Message>(msg);
+        m.ok()) {
+      push(sender, std::move(m.get()));
+    } else if (auto n = inspection::deserializeWithErrorT<ActorError>(msg);
+               n.ok()) {
+      push(sender, std::move(n.get()));
+    } else {
+      auto error =
+          ActorError{UnknownMessage{.sender = sender, .receiver = pid}};
+      auto payload = inspection::serializeWithErrorT(error);
+      if (payload.ok()) {
+        runtime->dispatch(pid, sender, payload.get());
+      } else {
+        fmt::print("Error serializing UnknownMessage");
+        std::abort();
+      }
+    }
+    kick();
```

Comment:
kick on every process is a bad idea, I already wrote about this
https://github.com/arangodb/arangodb/pull/17556#pullrequestreview-1181338093

But in last PR it wasn't devel branch, and I agreed it can be merged, in devel I don't think so

Reply:
fixed, kick is now only called when actor is idle